src/include/detail/linalg/tdb_io.h

```diff
@@ -376,63 +415,40 @@ auto sizes_to_indices(const std::vector<T>& sizes) {
  */
 template <class T>
 auto read_bin_local(const std::string& bin_file, size_t subset = 0) {
-  if (!std::filesystem::exists(bin_file)) {
-    throw std::runtime_error("file " + bin_file + " does not exist");
-  }
-  auto file_size = std::filesystem::file_size(bin_file);
-
-  auto fd = open(bin_file.c_str(), O_RDONLY);
-  if (fd == -1) {
-    throw std::runtime_error("could not open " + bin_file);
-  }
+  std::uint32_t dimension = 0u;
 
-  uint32_t dimension{0};
-  auto num_read = read(fd, &dimension, 4);
-  lseek(fd, 0, SEEK_SET);
+  const auto file_size = std::filesystem::file_size(bin_file);
+  auto file = std::ifstream{bin_file, std::ios::binary};
+  file.read(reinterpret_cast<char*>(&dimension), sizeof(dimension));
+  file.seekg(0);
 
-  auto max_vectors = file_size / (4 + dimension * sizeof(T));
+  const auto max_vectors = file_size / (4u + dimension * sizeof(T));
   if (subset > max_vectors) {
     throw std::runtime_error(
         "specified subset is too large " + std::to_string(subset) + " > " +
         std::to_string(max_vectors));
   }
-  auto num_vectors = subset == 0 ? max_vectors : subset;
 
-  struct stat s;
-  fstat(fd, &s);
-  size_t mapped_size = s.st_size;
-  assert(s.st_size == file_size);
+  const auto num_vectors = subset == 0 ? max_vectors : subset;
 
-  T* mapped_ptr = reinterpret_cast<T*>(
-      mmap(0, mapped_size, PROT_READ, MAP_FILE | MAP_PRIVATE, fd, 0));
+  auto result = ColMajorMatrix<T>{dimension, num_vectors};
+  auto* result_ptr = result.data();
 
-  if ((long)mapped_ptr == -1) {
-    throw std::runtime_error("mmap failed");
-  }
-
-  ColMajorMatrix<T> X(dimension, num_vectors);
-
-  auto data_ptr = X.data();
-  auto sift_ptr = mapped_ptr;
-
-  // Perform strided read
-  for (size_t k = 0; k < num_vectors; ++k) {
-    // Check for consistency of dimensions
-    decltype(dimension) dim = *reinterpret_cast<int*>(sift_ptr++);
-    if (dim != dimension) {
+  for (std::size_t i = 0; i < num_vectors; ++i) {
+    std::uint32_t d = 0u;
+    file.read(reinterpret_cast<char*>(&d), sizeof(d));
+    if (d != dimension) {
       throw std::runtime_error(
-          "dimension mismatch: " + std::to_string(dim) +
+          "dimension mismatch: " + std::to_string(d) +
           " != " + std::to_string(dimension));
     }
-    std::copy(sift_ptr, sift_ptr + dimension, data_ptr);
-    data_ptr += dimension;
-    sift_ptr += dimension;
+    if (not file.read(reinterpret_cast<char*>(result_ptr), d * sizeof(T))) {
```

Comment:
(Minor)(Optional) Does `clang-format` not fail on this? I would have expected:
```suggestion
    if (!file.read(reinterpret_cast<char*>(result_ptr), d * sizeof(T))) {
```

Reply:
Oh wait, it's failing. Will be curious to see once you run it.

Reply:
:( I am used to writing it like this, but my bad I will change it to `!` .